pkg/live/planner/cluster.go

```diff
@@ -0,0 +1,289 @@
+	if err != nil {
+		return nil, err
+	}
+
+	applier, err := apply.NewApplierBuilder().
+		WithFactory(f).
+		WithInventoryClient(invClient).
+		WithStatusPoller(statusPoller).
+		Build()
+	if err != nil {
+		return nil, err
+	}
+
+	return &ClusterPlanner{
+		applier:         applier,
+		resourceFetcher: fetcher,
+	}, nil
+}
+
+type Applier interface {
+	Run(ctx context.Context, invInfo inventory.Info, objects object.UnstructuredSet, options apply.ApplierOptions) <-chan event.Event
+}
+
+type ResourceFetcher interface {
+	FetchResource(ctx context.Context, id object.ObjMetadata) (*unstructured.Unstructured, bool, error)
+}
+
+type ClusterPlanner struct {
+	applier         Applier
+	resourceFetcher ResourceFetcher
+}
+
+type Options struct {
+	ServerSideOptions common.ServerSideOptions
+}
+
+func (r *ClusterPlanner) BuildPlan(ctx context.Context, inv inventory.Info, objects []*unstructured.Unstructured, o Options) (*planv1alpha1.Plan, error) {
+	actions, err := r.dryRunForPlan(ctx, inv, objects, o)
+	if err != nil {
+		return nil, err
+	}
+	return &planv1alpha1.Plan{
+		ResourceMeta: planv1alpha1.ResourceMeta,
+		Spec: planv1alpha1.PlanSpec{
+			Actions: actions,
+		},
+	}, nil
+}
+
+func (r *ClusterPlanner) dryRunForPlan(ctx context.Context, inv inventory.Info,
+	objects []*unstructured.Unstructured, o Options) ([]planv1alpha1.Action, error) {
+
+	eventCh := r.applier.Run(ctx, inv, objects, apply.ApplierOptions{
+		DryRunStrategy:    common.DryRunServer,
+		ServerSideOptions: o.ServerSideOptions,
+	})
+
+	var actions []planv1alpha1.Action
+	var err error
+	for e := range eventCh {
+		if e.Type == event.InitType {
+			// This event includes all resources that will be applied, pruned or deleted, so
+			// we make sure we fetch all the resources from the cluster.
+			// TODO: See if we can update the actuation library to provide the pre-actuation
+			// versions of the resources as part of the regular run. This solution is not great
+			// as fetching all resources will take time.
+			a, err := r.fetchResources(ctx, e)
+			if err != nil {
+				return nil, err
+			}
+			actions = a
+		}
+		if e.Type == event.ErrorType {
+			// Update the err variable here, but wait for the channel to close
+			// before we return from the function.
+			// Since ErrorEvents are considered fatal, there should only be sent
+			// and it will be followed by the channel being closed.
+			err = e.ErrorEvent.Err
+		}
+		// For the Apply, Prune and Delete event types, we just capture the result
+		// of the dry-run operation for the specific resource.
+		switch e.Type {
+		case event.ApplyType:
+			id := e.ApplyEvent.Identifier
+			index := indexForIdentifier(id, actions)
+			a := actions[index]
+			actions[index] = handleApplyEvent(e, a)
+		case event.PruneType:
+			id := e.PruneEvent.Identifier
+			index := indexForIdentifier(id, actions)
+			a := actions[index]
+			actions[index] = handlePruneEvent(e, a)
+		case event.DeleteType:
+			id := e.DeleteEvent.Identifier
```

Comment:
Q: What's the difference between a prune and a delete?

Reply:
Legacy behavior from the actuation library.